File: Backend/src/main/java/com/sencarmarket/module/certification/controller/CertificationController.java
package com.sencarmarket.module.certification.controller;

import com.sencarmarket.module.certification.dto.*;
import com.sencarmarket.module.certification.entity.DemandeCertification;
import com.sencarmarket.module.certification.service.CertificationService;
import com.sencarmarket.module.commun.dto.PaginatedResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/certifications")
@RequiredArgsConstructor
public class CertificationController {

    private final CertificationService certificationService;

    // ==================== Demande Certification ====================

    /**
     * Crée une nouvelle demande de certification
     */
    @PostMapping("/demandes")
    public ResponseEntity<DemandeCertificationResponse> createDemandeCertification(
            @Valid @RequestBody CreateDemandeCertificationRequest request,
            @RequestHeader("X-User-Id") UUID utilisateurId) {
        DemandeCertificationResponse response = certificationService.createDemandeCertification(request, utilisateurId);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    /**
     * Récupère toutes les demandes de certification avec pagination
     */
    @GetMapping("/demandes")
    public ResponseEntity<PaginatedResponse<DemandeCertificationResponse>> getAllDemandes(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        PaginatedResponse<DemandeCertificationResponse> response = certificationService.getAllDemandes(page, size);
        return ResponseEntity.ok(response);
    }

    /**
     * Récupère une demande de certification par ID
     */
    @GetMapping("/demandes/{id}")
    public ResponseEntity<DemandeCertificationResponse> getDemandeById(@PathVariable UUID id) {
        DemandeCertificationResponse response = certificationService.getDemandeById(id);
        return ResponseEntity.ok(response);
    }

    /**
     * Récupère les demandes de certification d'un utilisateur
     */
    @GetMapping("/demandes/utilisateur/{utilisateurId}")
    public ResponseEntity<List<DemandeCertificationResponse>> getDemandesByUtilisateur(
            @PathVariable UUID utilisateurId) {
        List<DemandeCertificationResponse> response = certificationService.getDemandesByUtilisateur(utilisateurId);
        return ResponseEntity.ok(response);
    }

    /**
     * Met à jour une demande de certification
     */
    @PutMapping("/demandes/{id}")
    public ResponseEntity<DemandeCertificationResponse> updateDemandeCertification(
            @PathVariable UUID id,
            @Valid @RequestBody CreateDemandeCertificationRequest request) {
        DemandeCertificationResponse response = certificationService.updateDemandeCertification(id, request);
        return ResponseEntity.ok(response);
    }

    /**
     * Supprime une demande de certification
     */
    @DeleteMapping("/demandes/{id}")
    public ResponseEntity<Void> deleteDemandeCertification(@PathVariable UUID id) {
        certificationService.deleteDemandeCertification(id);
        return ResponseEntity.noContent().build();
    }

    // ==================== Payment ====================

    /**
     * Traite le paiement d'une demande de certification
     */
    @PostMapping("/demandes/{demandeId}/payment")
    public ResponseEntity<DemandeCertificationResponse> processPayment(
            @PathVariable UUID demandeId,
            @RequestParam UUID paiementId) {
        DemandeCertificationResponse response = certificationService.processPayment(demandeId, paiementId);
        return ResponseEntity.ok(response);
    }

    // ==================== Inspector ====================

    /**
     * Attribue un inspecteur à une demande de certification
     */
    @PostMapping("/demandes/{demandeId}/assign-inspector")
    public ResponseEntity<DemandeCertificationResponse> assignInspector(
            @PathVariable UUID demandeId,
            @RequestParam UUID inspecteurId) {
        DemandeCertificationResponse response = certificationService.assignInspector(demandeId, inspecteurId);
        return ResponseEntity.ok(response);
    }

    // ==================== Statut ====================

    /**
     * Met à jour le statut d'une demande de certification
     */
    @PatchMapping("/demandes/{demandeId}/statut")
    public ResponseEntity<DemandeCertificationResponse> updateStatut(
            @PathVariable UUID demandeId,
            @RequestParam DemandeCertification.StatutDemande statut) {
        DemandeCertificationResponse response = certificationService.updateStatut(demandeId, statut);
        return ResponseEntity.ok(response);
    }

    // ==================== Inspection ====================

    /**
     * Crée une inspection pour une demande de certification
     */
    @PostMapping("/inspections")
    public ResponseEntity<InspectionResponse> createInspection(
            @Valid @RequestBody CreateInspectionRequest request,
            @RequestParam UUID demandeId) {
        InspectionResponse response = certificationService.createInspection(request, demandeId);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    /**
     * Récupère une inspection par ID
     */
    @GetMapping("/inspections/{id}")
    public ResponseEntity<InspectionResponse> getInspectionById(@PathVariable UUID id) {
        InspectionResponse response = certificationService.getInspectionById(id);
        return ResponseEntity.ok(response);
    }

    /**
     * Récupère les inspections d'un inspecteur avec pagination
     */
    @GetMapping("/inspections/inspecteur/{inspecteurId}")
    public ResponseEntity<PaginatedResponse<InspectionResponse>> getInspectionsByInspecteur(
            @PathVariable UUID inspecteurId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        PaginatedResponse<InspectionResponse> response = certificationService.getInspectionsByInspecteur(inspecteurId, page, size);
        return ResponseEntity.ok(response);
    }

    /**
     * Met à jour une inspection
     */
    @PutMapping("/inspections/{id}")
    public ResponseEntity<InspectionResponse> updateInspection(
            @PathVariable UUID id,
            @Valid @RequestBody CreateInspectionRequest request) {
        InspectionResponse response = certificationService.updateInspection(id, request);
        return ResponseEntity.ok(response);
    }

    /**
     * Supprime une inspection
     */
    @DeleteMapping("/inspections/{id}")
    public ResponseEntity<Void> deleteInspection(@PathVariable UUID id) {
        certificationService.deleteInspection(id);
        return ResponseEntity.noContent().build();
    }

    // ==================== Rapport ====================

    /**
     * Upload le PDF du rapport d'inspection
     */
    @PostMapping("/inspections/{inspectionId}/upload-rapport")
    public ResponseEntity<RapportInspectionResponse> uploadRapportPdf(
            @PathVariable UUID inspectionId,
            @RequestParam("file") MultipartFile file) {
        RapportInspectionResponse response = certificationService.uploadRapportPdf(inspectionId, file);
        return ResponseEntity.ok(response);
    }

    /**
     * Enregistre le résultat de l'inspection
     */
    @PostMapping("/inspections/{inspectionId}/resultat")
    public ResponseEntity<InspectionResponse> saveRapportResult(
            @PathVariable UUID inspectionId,
            @Valid @RequestBody CreateRapportInspectionRequest request) {
        InspectionResponse response = certificationService.saveRapportResult(inspectionId, request);
        return ResponseEntity.ok(response);
    }

    // ==================== Badge ====================

    /**
     * Génère le badge certifié pour un véhicule
     */
    @PostMapping("/demandes/{demandeId}/generate-badge")
    public ResponseEntity<DemandeCertificationResponse> generateBadge(@PathVariable UUID demandeId) {
        DemandeCertificationResponse response = certificationService.generateBadge(demandeId);
        return ResponseEntity.ok(response);
    }
}

File: Backend/src/modules/certification/certification.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseIntPipe,
  ParseUUIDPipe,
  Patch,
  Post,
  Put,
  Query,
  UploadedFile,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import { ApiBearerAuth, ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';

import { CurrentUser } from '../../common/decorators/current-user.decorator';
import { PaginatedResponseDto } from '../../common/dto/paginated-response.dto';
import { JwtAuthGuard } from '../../common/guards/jwt-auth.guard';
import type { AuthenticatedUser } from '../../common/types/authenticated-user.type';
import { ApiErrorResponseDto } from '../auth/dto/api-error-response.dto';

import { CertificationService } from './certification.service';
import { CreateDemandeCertificationRequestDto } from './dto/create-demande-certification-request.dto';
import { CreateInspectionRequestDto } from './dto/create-inspection-request.dto';
import { CreateRapportInspectionRequestDto } from './dto/create-rapport-inspection-request.dto';
import { DemandeCertificationResponseDto } from './dto/demande-certification-response.dto';
import { InspectionResponseDto } from './dto/inspection-response.dto';
import { RapportInspectionResponseDto } from './dto/rapport-inspection-response.dto';

type UploadedFileLike = { originalname?: string; buffer: Buffer };

@ApiTags('Certifications')
@UseGuards(JwtAuthGuard)
@ApiBearerAuth()
@Controller('certifications')
export class CertificationController {
  constructor(private readonly service: CertificationService) {}

  @Post('demandes')
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: 'Créer une demande de certification' })
  @ApiResponse({ status: 201, type: DemandeCertificationResponseDto, description: 'Demande créée avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  createDemandeCertification(
    @Body() request: CreateDemandeCertificationRequestDto,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto> {
    return this.service.createDemandeCertification(request, user);
  }

  @Get('demandes')
  @ApiOperation({ summary: 'Liste des demandes de certification' })
  @ApiResponse({ status: 200, description: 'Demandes récupérées avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getAllDemandes(
    @CurrentUser() user: AuthenticatedUser,
    @Query('page', new ParseIntPipe({ optional: true })) page = 0,
    @Query('size', new ParseIntPipe({ optional: true })) size = 10,
  ): Promise<PaginatedResponseDto<DemandeCertificationResponseDto>> {
    return this.service.getAllDemandes(page, size, user);
  }

  @Get('demandes/utilisateur/:utilisateurId')
  @ApiOperation({ summary: 'Demandes de certification d\'un utilisateur' })
  @ApiResponse({ status: 200, description: 'Demandes récupérées avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Utilisateur non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getDemandesByUtilisateur(
    @Param('utilisateurId', new ParseUUIDPipe()) utilisateurId: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto[]> {
    return this.service.getDemandesByUtilisateur(utilisateurId, user);
  }

  @Get('demandes/:id')
  @ApiOperation({ summary: 'Obtenir une demande de certification par ID' })
  @ApiResponse({ status: 200, type: DemandeCertificationResponseDto, description: 'Demande récupérée avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getDemandeById(
    @Param('id', new ParseUUIDPipe()) id: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto> {
    return this.service.getDemandeById(id, user);
  }

  @Put('demandes/:id')
  @ApiOperation({ summary: 'Mettre à jour une demande de certification' })
  @ApiResponse({ status: 200, type: DemandeCertificationResponseDto, description: 'Demande mise à jour avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  updateDemandeCertification(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() request: CreateDemandeCertificationRequestDto,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto> {
    return this.service.updateDemandeCertification(id, request, user);
  }

  @Delete('demandes/:id')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Supprimer une demande de certification' })
  @ApiResponse({ status: 204, description: 'Demande supprimée avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  async deleteDemandeCertification(
    @Param('id', new ParseUUIDPipe()) id: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<void> {
    await this.service.deleteDemandeCertification(id, user);
  }

  @Post('demandes/:demandeId/payment')
  @ApiOperation({ summary: 'Traiter le paiement d\'une demande de certification' })
  @ApiResponse({ status: 200, type: DemandeCertificationResponseDto, description: 'Paiement traité avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Paiement invalide ou échoué' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  processPayment(
    @Param('demandeId', new ParseUUIDPipe()) demandeId: string,
    @Query('paiementId', new ParseUUIDPipe()) paiementId: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto> {
    return this.service.processPayment(demandeId, paiementId, user);
  }

  @Post('demandes/:demandeId/assign-inspector')
  @ApiOperation({ summary: 'Attribuer un inspecteur' })
  @ApiResponse({ status: 200, type: DemandeCertificationResponseDto, description: 'Inspecteur attribué avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  assignInspector(
    @Param('demandeId', new ParseUUIDPipe()) demandeId: string,
    @Query('inspecteurId', new ParseUUIDPipe()) inspecteurId: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto> {
    return this.service.assignInspector(demandeId, inspecteurId, user);
  }

  @Patch('demandes/:demandeId/statut')
  @ApiOperation({ summary: 'Mettre à jour le statut d\'une demande de certification' })
  @ApiResponse({ status: 200, type: DemandeCertificationResponseDto, description: 'Statut mis à jour avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Statut invalide' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  updateStatut(
    @Param('demandeId', new ParseUUIDPipe()) demandeId: string,
    @Query('statut') statut: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto> {
    return this.service.updateStatutFromRaw(demandeId, statut, user);
  }

  @Post('demandes/:demandeId/inspections')
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: 'Créer une inspection' })
  @ApiResponse({ status: 201, type: InspectionResponseDto, description: 'Inspection créée avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  createInspection(
    @Body() request: CreateInspectionRequestDto,
    @Param('demandeId', new ParseUUIDPipe()) demandeId: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<InspectionResponseDto> {
    return this.service.createInspection(request, demandeId, user);
  }

  @Get('inspections/:id')
  @ApiOperation({ summary: 'Obtenir une inspection par ID' })
  @ApiResponse({ status: 200, type: InspectionResponseDto, description: 'Inspection récupérée avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Inspection non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getInspectionById(
    @Param('id', new ParseUUIDPipe()) id: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<InspectionResponseDto> {
    return this.service.getInspectionById(id, user);
  }

  @Get('inspections/inspecteur/:inspecteurId')
  @ApiOperation({ summary: 'Inspections d\'un inspecteur' })
  @ApiResponse({ status: 200, description: 'Inspections récupérées avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Inspecteur non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getInspectionsByInspecteur(
    @Param('inspecteurId', new ParseUUIDPipe()) inspecteurId: string,
    @Query('page', new ParseIntPipe({ optional: true })) page = 0,
    @Query('size', new ParseIntPipe({ optional: true })) size = 10,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<PaginatedResponseDto<InspectionResponseDto>> {
    return this.service.getInspectionsByInspecteur(inspecteurId, page, size, user);
  }

  @Put('inspections/:id')
  @ApiOperation({ summary: 'Mettre à jour une inspection' })
  @ApiResponse({ status: 200, type: InspectionResponseDto, description: 'Inspection mise à jour avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Inspection non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  updateInspection(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() request: CreateInspectionRequestDto,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<InspectionResponseDto> {
    return this.service.updateInspection(id, request, user);
  }

  @Delete('inspections/:id')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Supprimer une inspection' })
  @ApiResponse({ status: 204, description: 'Inspection supprimée avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Inspection non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  async deleteInspection(
    @Param('id', new ParseUUIDPipe()) id: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<void> {
    await this.service.deleteInspection(id, user);
  }

  @Post('inspections/:inspectionId/upload-rapport')
  @UseInterceptors(FileInterceptor('file'))
  @ApiOperation({ summary: 'Uploader le PDF du rapport d\'inspection' })
  @ApiResponse({ status: 200, type: RapportInspectionResponseDto, description: 'Rapport uploadé avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Fichier invalide' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Inspection non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  uploadRapportPdf(
    @Param('inspectionId', new ParseUUIDPipe()) inspectionId: string,
    @UploadedFile() file: UploadedFileLike | undefined,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<RapportInspectionResponseDto> {
    return this.service.uploadRapportPdf(inspectionId, file as UploadedFileLike, user);
  }

  @Post('inspections/:inspectionId/resultat')
  @ApiOperation({ summary: 'Enregistrer le résultat du rapport d\'inspection' })
  @ApiResponse({ status: 200, type: InspectionResponseDto, description: 'Résultat enregistré avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Inspection non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  saveRapportResult(
    @Param('inspectionId', new ParseUUIDPipe()) inspectionId: string,
    @Body() request: CreateRapportInspectionRequestDto,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<InspectionResponseDto> {
    return this.service.saveRapportResult(inspectionId, request, user);
  }

  @Post('demandes/:demandeId/generate-badge')
  @ApiOperation({ summary: 'Générer le badge certifié' })
  @ApiResponse({ status: 200, type: DemandeCertificationResponseDto, description: 'Badge généré avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Demande non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  generateBadge(
    @Param('demandeId', new ParseUUIDPipe()) demandeId: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<DemandeCertificationResponseDto> {
    return this.service.generateBadge(demandeId, user);
  }
}
